src/net_processing.cpp

```diff
@@ -731,6 +735,37 @@ std::chrono::microseconds CalculateTxGetDataTime(const uint256& txid, std::chron
     return process_time;
 }
 
+static void RetryProcessTx(CConnman& connman, const uint256& txid, const std::chrono::microseconds current_time) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    CNodeState::TxDownloadState* best_d = nullptr;
+    std::chrono::microseconds best;
+
+    for (auto& el : mapNodeState) {
+         CNodeState::TxDownloadState* d = &el.second.m_tx_download;
+         if (d->m_tx_in_flight.size() >= MAX_PEER_TX_IN_FLIGHT) continue;
+         auto it = d->m_tx_announced.find(txid);
+         if (it != d->m_tx_announced.end()) {
+             if (best_d == nullptr || (it->second != std::chrono::microseconds::zero() && it->second < best)) {
+                 best_d = d;
+                 best = it->second;
+             }
+         }
+    }
+
+    std::chrono::microseconds process_time = current_time + GetRandMicros(MAX_NOTFOUND_RETRY_RANDOM_DELAY);
```

Comment:
Why is 2-seconds random delay useful here? It can accidentally prioritize inbound nodes over outbound :)

Reply:
It already tests that `process_time < best` so the 2s delay will only bring a process time forward, not push it back past an inbound peer's process time.

I think the scenario is likely to be:

 * bunch of peers announce a new tx, outbound connections get polled asap, inbound connections get polled after 2s
 * you pick one of those, likely an outbound, and send a GETDATA
 * other peers announce, get a 60s-64s delay
 * the early announcing peers get their process time rescheduled with an additional 60-64s delay
 * a NOTFOUND comes back, and you pick whoever's delay was shortest; if it came back in under 2s, you'll next ask an inbound connection who announced early; if it came back later, you'll probably ask an outbound connection who announced after you'd sent the first GETDATA

Which is to say, I don't think we're consistently prioritising outbounds after the first request anyway.

I don't have a good rationale for the 2s random delay; instantly going from a NOTFOUND from one peer to a GETDATA to another peer seems risky to me though.

Reply:
Yeah, I was a bit confused, now what's on your mind makes sense to me.
I'm probably fine with leaving your 2s random delay.

>Which is to say, I don't think we're consistently prioritising outbounds after the first request anyway.

I think we do? That's probably off-topic here, although i'd be curious why you think so.